File: Codigos/MedianOfTwoSortedArrays.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <algorithm>

class Solution {
public:
    double findMedianSortedArrays(std::vector<int>& nums1, std::vector<int>& nums2) 
    {
        for (const auto& x : nums2)
            nums1.push_back(x);

        std::sort(nums1.begin(), nums1.end());

        if (nums1.size() % 2 == 0) // tamanho par
        {
            int meio = static_cast<int>(nums1.size()) / 2;
            return static_cast<double>(nums1[meio] + nums1[meio - 1]) / 2;
        }
        else //Tamanho impar
        {
            double meio = static_cast<double>(nums1.size()) / 2;
            meio -= std::fmod(meio, 1);
            return nums1[meio];
        }
    }
};

int main()
{
    std::vector<int> nums1{ 1, 2 };
    std::vector<int> nums2{ 3, 4, 5 };
    Solution x;

    std::cout << x.findMedianSortedArrays(nums1, nums2);
}

File: Codigos Python/Median_of_Two_Sorted_Arrays.py
# Fonte: https://leetcode.com/problems/median-of-two-sorted-arrays/
# Autor: Bruno Harlis
# Data: 29/07/2021
"""
PROBLEMA PROPOSTO:
Dadas duas matrizes classificadas nums1 e nums2 de tamanho m e n respectivamente, 
retorna a mediana das duas matrizes classificadas.

A complexidade geral do tempo de execução deve ser O(log (m+n)).

Exemplo 1:
Entrada: nums1 = [1,3], nums2 = [2]
Saída: 2,00000
Explicação: matriz mesclada = [1,2,3] e a mediana é 2.

Tempo de execução : 88 ms, mais rápido do que 84,93 % dos envios online do Python3.
Uso de Memória : 14,4 MB, menos de 81,38 % de python3.
"""


class Solution:
    def findMedianSortedArrays(self, nums1: List[int], nums2: List[int]) -> float:
        for i in nums2:
        nums1.append(i)

        nums1.sort()

        par = True
        if len(nums1) % 2 != 0:
            par = False

        meio = len(nums1) // 2

        if par:
            return (nums1[meio] + nums1[meio - 1]) / 2
        else:
            return nums1[meio]
